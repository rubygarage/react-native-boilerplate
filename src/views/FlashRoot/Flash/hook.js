import { useRef, useEffect } from 'react';
import { useDispatch } from 'react-redux';

import { hideFlash } from 'state/flash/actions';

const useContainer = (props) => {
  const timerHandle = useRef();

  const dispatch = useDispatch();

  const {
    id,
    flashProps: { action },
    doAnimation,
    lifetime,
  } = props;

  const onHideFlash = () => {
    dispatch(hideFlash(id));
    doAnimation();
  };

  const onActionClick = () => {
    dispatch(hideFlash(id));
    dispatch(action);
  };

  const didMount = () => {
    if (lifetime) {
      timerHandle.current = setInterval(onHideFlash, lifetime);
    }

    return () => {
      clearInterval(timerHandle.current);
    };
  };

  useEffect(didMount, []);

  return {
    onHideFlash,
    onActionClick,
    didMount,
  };
};

export default useContainer;
