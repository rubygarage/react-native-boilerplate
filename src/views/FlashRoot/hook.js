import { useEffect } from 'react';
import { useSelector } from 'react-redux';
import { UIManager, LayoutAnimation } from 'react-native';

import { flashesSelector } from 'state/flash/selectors';

export const ANIMATION_DURATION = 250;

const useContainer = () => {
  const flashes = useSelector(flashesSelector);

  const doAnimation = () => {
    LayoutAnimation.configureNext({
      duration: ANIMATION_DURATION,
      update: {
        type: LayoutAnimation.Types.linear,
      },
    });
    LayoutAnimation.configureNext({
      duration: ANIMATION_DURATION,
      create: {
        type: LayoutAnimation.Types.linear,
        property: LayoutAnimation.Properties.scaleXY,
      },
    });
  };

  useEffect(() => {
    if (UIManager.setLayoutAnimationEnabledExperimental) {
      UIManager.setLayoutAnimationEnabledExperimental(true);
    }
  }, []);

  useEffect(doAnimation, [flashes.length]);

  return {
    doAnimation,
  };
};

export default useContainer;
