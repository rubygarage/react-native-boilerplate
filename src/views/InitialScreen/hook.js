import { useEffect } from 'react';
import SplashScreen from 'react-native-splash-screen';
import { useNavigation } from '@react-navigation/native';

const useContainer = () => {
  const navigation = useNavigation();

  /**
   * Mounting
   */
  const didMount = () => {
    SplashScreen.hide();
    navigation.navigate('Home');
  };

  useEffect(didMount, []);

  return {
    navigation,
    didMount,
  };
};

export default useContainer;
