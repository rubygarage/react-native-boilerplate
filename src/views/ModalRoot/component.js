import React from 'react';

import MODAL_COMPONENTS from './modalComponents';
import useContainer from './hook';

const ModalRoot = () => {
  const {
    modalType,
    modalProps,
    isVisible,
    onCloseModal,
    onDestroyModal,
  } = useContainer();

  if (!modalType) {
    return null;
  }

  const SpecificModal = MODAL_COMPONENTS[modalType].modal;

  return (
    <SpecificModal
      isVisible={isVisible}
      onCloseModal={onCloseModal}
      onDestroyModal={onDestroyModal}
      {...modalProps}
    />
  );
};

export default ModalRoot;
