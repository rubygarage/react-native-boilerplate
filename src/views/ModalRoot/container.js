import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { hideModal } from 'state/modal/actions';
import ModalRootComponent from './component';

class ModalRoot extends Component {
  state = { isVisible: true }

  onCloseModal = () => {
    this.setState({ isVisible: false });
  }

  onDestroyModal = () => {
    const { destroyModal } = this.props;
    destroyModal();
    this.setState({ isVisible: true });
  }

  render() {
    const { isVisible } = this.state;
    return (
      <ModalRootComponent
        {...this.props}
        onCloseModal={this.onCloseModal}
        onDestroyModal={this.onDestroyModal}
        isVisible={isVisible}
      />
    );
  }
}

ModalRoot.propTypes = {
  destroyModal: PropTypes.func.isRequired,
};

const mapStateToProps = (state) => ({
  modalType: state.modal.modalType,
  modalProps: state.modal.modalProps,
});

const mapDispatchToProps = {
  destroyModal: hideModal,
};

export { ModalRoot as ModalRootContainer };
export default connect(mapStateToProps, mapDispatchToProps)(ModalRoot);
