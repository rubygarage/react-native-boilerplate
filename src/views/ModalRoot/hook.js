import { useState } from 'react';
import { useSelector, useDispatch } from 'react-redux';

import { modalTypeSelector, modalPropsSelector } from 'state/modal/selectors';
import { hideModal } from 'state/modal/actions';

const useContainer = () => {
  const modalType = useSelector(modalTypeSelector);
  const modalProps = useSelector(modalPropsSelector);

  const dispatch = useDispatch();

  const [isVisible, setIsVisible] = useState(true);

  const onCloseModal = () => {
    setIsVisible(false);
  };

  const onDestroyModal = () => {
    dispatch(hideModal());
    setIsVisible(true);
  };

  return {
    modalType,
    modalProps,
    isVisible,
    setIsVisible,
    onCloseModal,
    onDestroyModal,
  };
};

export default useContainer;
