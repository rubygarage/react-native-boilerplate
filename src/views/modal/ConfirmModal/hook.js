import { useContext, useState } from 'react';
import { ThemeContext } from 'styled-components';
import { useDispatch } from 'react-redux';

const useContainer = (props) => {
  const theme = useContext(ThemeContext);

  const { onCloseModal, onDestroyModal, confirmAction } = props;

  const dispatch = useDispatch();

  const [isShouldConfirm, setShouldConfirm] = useState(false);

  const handleShouldConfirm = () => {
    setShouldConfirm(true);
    onCloseModal();
  };

  const onModalHide = () => {
    if (isShouldConfirm) {
      dispatch(confirmAction);
    }
    onDestroyModal();
  };

  return {
    theme,
    handleShouldConfirm,
    onModalHide,
    setShouldConfirm,
    isShouldConfirm,
  };
};

export default useContainer;
