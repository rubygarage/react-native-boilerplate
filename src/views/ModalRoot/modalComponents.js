import ConfirmModal from 'views/modal/ConfirmModal';
import InfoModal from 'views/modal/InfoModal';

const MODAL_COMPONENTS = {
  CONFIRM_MODAL: { key: 'CONFIRM_MODAL', modal: ConfirmModal },
  INFO_MODAL: { key: 'INFO_MODAL', modal: InfoModal },
};

export default MODAL_COMPONENTS;
