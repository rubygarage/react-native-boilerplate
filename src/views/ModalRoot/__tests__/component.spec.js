import React from 'react';
import { shallow } from 'enzyme';

import ModalRoot from '../component';
import * as useContainer from '../hook';

jest.mock('../modalComponents', () => ({ MY_MODAL: { key: 'MY_MODAL', modal: jest.fn() } }));

jest.mock('../hook', () => ({
  __esModule: true,
  default: jest.fn(() => ({
    modalType: 'MY_MODAL',
    modalProps: { customProp: 'customProp' },
    isVisible: true,
    onCloseModal: jest.fn(),
    onDestroyModal: jest.fn(),
  })),
}));

describe('ModalRoot component', () => {
  it('renders ModalRoot when modalType is present', () => {
    const component = shallow(<ModalRoot />);

    expect(component).toMatchSnapshot();
  });

  it('return null unless modalType present', () => {
    useContainer.default.mockImplementation(() => ({
      modalType: null,
      modalProps: {},
      isVisible: true,
      onCloseModal: jest.fn(),
      onDestroyModal: jest.fn(),
    }));

    const component = shallow(<ModalRoot />);

    expect(component).toMatchSnapshot();
  });
});
