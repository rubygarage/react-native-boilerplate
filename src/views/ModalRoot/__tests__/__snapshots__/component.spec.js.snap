// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalRoot component renders ModalRoot when modalType is present 1`] = `
<mockConstructor
  customProp="customProp"
  isVisible={true}
  onCloseModal={[MockFunction]}
  onDestroyModal={[MockFunction]}
/>
`;

exports[`ModalRoot component return null unless modalType present 1`] = `""`;
